src/api-service/__app__/onefuzzlib/job_templates/render.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import json
+from typing import Dict, List
+
+from jsonpatch import apply_patch
+from memoization import cached
+from onefuzztypes.enums import ContainerType, UserFieldType
+from onefuzztypes.job_templates import (
+    TEMPLATE_BASE_FIELDS,
+    JobTemplate,
+    JobTemplateConfig,
+    JobTemplateField,
+    JobTemplateRequest,
+    TemplateUserData,
+    UserField,
+)
+
+
+def template_container_types(template: JobTemplate) -> List[ContainerType]:
+    return list(set(y.type for x in template.tasks for y in x.containers if not y.name))
+
+
+@cached
+def build_input_config(name: str, template: JobTemplate) -> JobTemplateConfig:
+    user_fields = [
+        JobTemplateField(
+            name=x.name,
+            type=x.type,
+            required=x.required,
+            default=x.default,
+            help=x.help,
+        )
+        for x in TEMPLATE_BASE_FIELDS + template.user_fields
+    ]
+    containers = template_container_types(template)
+
+    return JobTemplateConfig(
+        name=name,
+        user_fields=user_fields,
+        containers=containers,
+    )
+
+
+def build_patches(
+    data: TemplateUserData, field: UserField
+) -> List[Dict[str, TemplateUserData]]:
+    patches = []
+
+    if field.type == UserFieldType.Bool and not isinstance(data, bool):
+        raise Exception("invalid bool field")
+    if field.type == UserFieldType.Int and not isinstance(data, int):
+        raise Exception("invalid int field")
+    if field.type == UserFieldType.Str and not isinstance(data, str):
+        raise Exception("invalid str field")
+    if field.type == UserFieldType.DictStr and not isinstance(data, dict):
+        raise Exception("invalid DictStr field")
+    if field.type == UserFieldType.ListStr and not isinstance(data, list):
+        raise Exception("invalid ListStr field")
+
+    for location in field.locations:
+        patches.append(
+            {
+                "op": location.op.name,
+                "path": location.path,
+                "value": data,
+            }
+        )
+
+    return patches
+
+
+def render(request: JobTemplateRequest, template: JobTemplate) -> JobTemplate:
+    patches = []
+    seen = set()
+
+    for name in request.user_fields:
+        for field in TEMPLATE_BASE_FIELDS + template.user_fields:
+            if field.name == name:
+                if name in seen:
+                    raise ValueError(f"duplicate specification: {name}")
+                seen.add(name)
+
+        if name not in seen:
```

Comment:
Not sure if it's worth throwing an error in this case. allowing the template to be over specified could make it easier for client to execute multiple templates with the same set of parameters

Reply:
With my suggestion of returning the error as part of the return type this can just be a warning instead

Reply:
A few issues with that idea.  

1. Warnings at this level only go to application insights.
2. Warnings would only hit after the jobs are submitted.  If it was because of a typo, users could accidentally do something catastrophic.  I'd rather avoid that.

However, instead of generating an exception, these should be turned into errors which will get sent back to the user.